### PR TITLE
Updating to pytorch V.4, adding device instead of Cuda for API

### PR DESCRIPTION
**Pull request by renebidart**
Updating API to pytorch V.4, using .to(device) instead of .cuda()
Change the cuda argument we could to device

### CODE REVIEW

File: foolbox/models/pytorch.py [hunk 1]
```diff
@@ -163,24 +157,20 @@ def _loss_fn(self, image, label):
...
         if self._old_pytorch():  # pragma: no cover
             images = Variable(images, volatile=True)
         predictions = self._model(images)
         ce = nn.CrossEntropyLoss()
         loss = ce(predictions, target)
         if self._old_pytorch():  # pragma: no cover
             loss = loss.data
-        if self.cuda:  # pragma: no cover
-            loss = loss.cpu()
+        if self.device != "cpu":
```
Comment (wielandbrendel) [inline foolbox/models/pytorch.py]: If a tensor is already on cpu does to(device) carry any overhead? I guess it does not, in which case we could remove the check for cpu here and in other places of the script.

Comment (renebidart) [reply foolbox/models/pytorch.py]: Ok, I'll change it to always put it on the correct device.

File: foolbox/models/pytorch.py [hunk 2]
```diff
@@ -108,14 +105,10 @@ def predictions_and_gradient(self, image, label):
...
-            target = target.cuda()
+        target = torch.from_numpy(target).to(self.device)
 
         images = image[np.newaxis]
-        images = torch.from_numpy(images)
-        if self.cuda:  # pragma: no cover
-            images = images.cuda()
+        images = torch.from_numpy(images).to(self.device)
 
         if self._old_pytorch():  # pragma: no cover
```
Comment (wielandbrendel) [inline foolbox/models/pytorch.py]: By using .to(device) we are anyway breaking compatibility with older pytorch versions in which case we can remove these checks and accompanying fallbacks altogether.

**Comment by wielandbrendel** [comment]
@jonasrauber Are you fine with removing PyTorch v0.3 and lower support? Most PyTorch users I know have very quickly jumped to 0.4 so I'd think this should be fine.

File: foolbox/models/pytorch.py [hunk 3]
```diff
@@ -18,14 +18,13 @@ class PyTorchModel(DifferentiableModel):
         Number of classes for which the model will output predictions.
     channel_axis : int
         The index of the axis that represents color channels.
-    cuda : bool
-        A boolean specifying whether the model uses CUDA. If None,
-        will default to torch.cuda.is_available()
+    device : string
+        A string specifying the device to do computation on.
```
Comment (jonasrauber) [inline foolbox/models/pytorch.py]: Passing a string or a `torch.device` should both be supported and documented.

**Comment by jonasrauber** [comment]
As @wielandbrendel mentioned: tests need to be updated to use the device argument (or actually, maybe they do not need to pass this at all anymore; just leaving it at None migth be sufficient for the tests).

Some other thoughts:

I am not sure whether we should drop PyTorch 0.3 support. In principle I am in favor of it, but are we sure they are no important legacy models implemented only in PyTorch 0.3 that need to be tested?

Regarding backwards compatibility: I am wondering whether we should accept the `cuda` argument in addition to the `device` argument, to avoid breaking existing code using Foolbox. Not sure yet whether this is important enough to add this complexity to our code given that it should be easy to fix this problem if someone runs into this.

**Comment by jonasrauber** [comment]
will solve #195

**Comment by wielandbrendel** [comment]
I'd say we can cut Pytorch 0.3 support. Even if some legacy models are still around it is fairly easy to update them to 0.4. And in the worst case people can still use a slightly older version of Foolbox.

**Comment by jonasrauber** [comment]
Agreed, let's drop PyTorch 0.3 :)
Still wondering whether we should keep the `cuda` argument for Foolbox API backwards compatibility, but I think we can also drop that.
Let's fix the tests and then we should be close to merging. (I will fix them at some point if no one else does, @renebidart ?)

**Comment by renebidart** [comment]
@jonasrauber by fixing the tests do you mean updating this https://github.com/bethgelab/foolbox/blob/92e3cec74b63b84d03275d30a2932f408aaf8269/foolbox/tests/conftest.py#L114
to include the device argument instead of cuda? Or is there more to it?

**Comment by jonasrauber** [comment]
@renebidart That should probably  be it, but in the end, the automatic tests run by travis have to pass (i.e. that's the ultimate judge).

**Comment by jonasrauber** [comment]
There is still one bug: self. missing

**Comment by jonasrauber** [comment]
@renebidart I had some trouble fixing your pull-request because you made it from your master branch (you should be using separate branches for PRs)… your PR can now be found here: #206

**Comment by renebidart** [comment]
Sorry about that, next time I'll do it properly.

**Pull request closed without merging**